### PR TITLE
[SDCT-837] Document new Test Impact Analysis limitations

### PR DESCRIPTION
**Pull request by TovRudyy**
### What does this PR do? What is the motivation?

We want to release a new, mor efficient, implementation of Test Impact Analysis which introduces new limitations and unexpected behaviour. This PR tries to document the new changes that may impact customers.

### Merge readiness

- [ ] Ready for merge (**do not merge immediately**)

### CODE REVIEW

File: content/en/tests/test_impact_analysis/troubleshooting/_index.md [hunk 1]
```diff
@@ -33,6 +33,19 @@ For large applications with more than 50,000 tests, Datadog may return fewer ski
 export DD_TESTOPTIMIZATION_TIA_TEST_SKIPPING_MODE=suite
 {{< /code-block >}}
```
Comment (TovRudyy) [inline content/en/tests/test_impact_analysis/troubleshooting/_index.md]: remove this text, it is no longer true

File: content/en/tests/test_impact_analysis/troubleshooting/_index.md [hunk 2]
```diff
@@ -33,6 +33,19 @@ For large applications with more than 50,000 tests, Datadog may return fewer ski
...
 
+Other situations can also lead to fewer skipped tests. In all of them the extra tests run safely: Test Impact Analysis never skips a test that should have run.
```
Comment (TovRudyy) [inline content/en/tests/test_impact_analysis/troubleshooting/_index.md]:
```suggestion
There are edge-cases were Test Impact Analysis will skip less tests than expected.
```

File: content/en/tests/test_impact_analysis/troubleshooting/_index.md [hunk 3]
```diff
@@ -33,6 +33,19 @@ For large applications with more than 50,000 tests, Datadog may return fewer ski
...
+
+- A very large changeset may result in slightly fewer tests being skipped, because Datadog analyzes up to the 100 most recent commits.
```
Comment (TovRudyy) [inline content/en/tests/test_impact_analysis/troubleshooting/_index.md]:
```suggestion
- A very large git history may result in slightly fewer tests being skipped, because Datadog analyzes up to the first 100 recent commits.
```

File: content/en/tests/test_impact_analysis/troubleshooting/_index.md [hunk 4]
```diff
@@ -33,6 +33,19 @@ For large applications with more than 50,000 tests, Datadog may return fewer ski
...
+- A single commit that changes more than 5,000 files is not analyzed for impact, so its tests run rather than being skipped.
```
Comment (TovRudyy) [inline content/en/tests/test_impact_analysis/troubleshooting/_index.md]:
```suggestion
- A commit that changed more than 5,000 files is not analyzed for impact, so it is expected the coverage information of that commit to be ignored.
```

File: content/en/tests/test_impact_analysis/troubleshooting/_index.md [hunk 5]
```diff
@@ -33,6 +33,19 @@ For large applications with more than 50,000 tests, Datadog may return fewer ski
...
+- A test or suite that covers more than 16,000 files is not skipped.
```
Comment (TovRudyy) [inline content/en/tests/test_impact_analysis/troubleshooting/_index.md]:
```suggestion
- A test or suite that covers more than 16,000 files is never skipped.
```

File: content/en/tests/test_impact_analysis/troubleshooting/_index.md [hunk 6]
```diff
@@ -33,6 +33,19 @@ For large applications with more than 50,000 tests, Datadog may return fewer ski
...
+- A very small fraction of tests (well under one percent) may run even when they could have been skipped.
+
+For more detail on these behaviors, see [Limitations and expected behavior][7].
+
+### Code coverage is incomplete or missing
+
+Per-test (line-level) code coverage is collected for Java, JavaScript, .NET, and Go. It is not collected for Ruby, Python, or Swift. For very large test suites, per-test coverage may occasionally be unavailable even in supported languages, and affected tests are marked with a missing-coverage indicator. Suite-level coverage remains reliable and is what Test Impact Analysis uses to decide which tests to skip.
```
Comment (TovRudyy) [inline content/en/tests/test_impact_analysis/troubleshooting/_index.md]: This is all wrong. First of all, this should be inside the "Code Coverage Backfilling" section. Second, the feature that may be degraded is the code coverage backfill (i.e.: backfill the test coverage of tests skipped by TIA in the code coverage report of the run)

File: content/en/tests/test_impact_analysis/_index.md [hunk 7]
```diff
@@ -49,6 +49,32 @@ There are several configuration mechanisms that you can use in these scenarios t
...
+
+<div class="alert alert-info">All of the behaviors in this section are safe by design: they can only cause <strong>more</strong> tests to run, never cause a test to be skipped when it should have run. For code changes that Datadog cannot detect automatically, see <a href="#out-of-the-box-configuration-limitations">Out-of-the-box configuration limitations</a>.</div>
+
+### Code coverage
+
+Test Impact Analysis relies on code coverage collected from previous test runs. Suite-level coverage is reliable and is what Test Impact Analysis uses to decide which tests to skip. For very large test suites, per-test (line-level) coverage may occasionally be unavailable. When it is, Datadog marks the affected test with a missing-coverage indicator, so you can tell a test with no covered lines apart from one whose per-test coverage was not recorded.
+
+Per-test code coverage is collected for Java, JavaScript, .NET, and Go. It is not collected for Ruby, Python, or Swift. See [Code coverage backfilling](#code-coverage-backfilling) for how tests skipped by Test Impact Analysis are accounted for in your total reported coverage.
```
Comment (TovRudyy) [inline content/en/tests/test_impact_analysis/_index.md]: remove this

**Review by anmarchenko — APPROVED** [review]

File: content/en/tests/test_impact_analysis/troubleshooting/_index.md [hunk 8]
```diff
@@ -27,11 +27,14 @@ Test Impact Analysis only takes into account the commit history and test code co
 
 ### Fewer tests skipped than expected
 
-For large applications with more than 50,000 tests, Datadog may return fewer skippable tests than expected. If you use Ruby and expect more tests to be skipped, switch to suite-level skipping by setting this environment variable before running tests:
+There are edge-cases where Test Impact Analysis will skip less tests than expected.
```
Comment (OliviaShoup) [inline content/en/tests/test_impact_analysis/troubleshooting/_index.md]: Three issues in one sentence:

- `edge-cases` -> `edge cases` (not hyphenated as a noun)
- `will skip` -> present tense; Vale flags `will` as a temporal word (`Datadog.tense`)
- `less tests` -> `fewer tests` ("tests" is a count noun)

```suggestion
There are edge cases where Test Impact Analysis skips fewer tests than expected.
```

File: content/en/tests/test_impact_analysis/troubleshooting/_index.md [hunk 9]
```diff
@@ -27,11 +27,14 @@ Test Impact Analysis only takes into account the commit history and test code co
...
 
-{{< code-block lang="bash" >}}
-export DD_TESTOPTIMIZATION_TIA_TEST_SKIPPING_MODE=suite
-{{< /code-block >}}
+- A very large git history may result in slightly fewer tests being skipped, because Datadog analyzes up to the first 100 recent commits.
```
Comment (OliviaShoup) [inline content/en/tests/test_impact_analysis/troubleshooting/_index.md]: "the first 100 recent commits" is self-contradictory — "first" and "recent" pull in opposite directions. The main page has it right ("the 100 most recent commits"). Also `git` -> `Git` per the word list (capitalized except in commands).

```suggestion
- A very large Git history may result in slightly fewer tests being skipped, because Datadog analyzes up to the 100 most recent commits.
```

File: content/en/tests/test_impact_analysis/troubleshooting/_index.md [hunk 10]
```diff
@@ -27,11 +27,14 @@ Test Impact Analysis only takes into account the commit history and test code co
...
+- A commit that changed more than 5,000 files is not analyzed for impact, so it is expected the coverage information of that commit to be ignored.
```
Comment (OliviaShoup) [inline content/en/tests/test_impact_analysis/troubleshooting/_index.md]: "it is expected the coverage information ... to be ignored" isn't a valid construction, and the hedge adds nothing — the information *is* ignored. The tense also doesn't match the sibling bullets ("changed" here vs. present tense elsewhere).

```suggestion
- A commit that changes more than 5,000 files is not analyzed for impact, so the coverage information from that commit is ignored.
```

File: content/en/tests/test_impact_analysis/troubleshooting/_index.md [hunk 11]
```diff
@@ -27,11 +27,14 @@ Test Impact Analysis only takes into account the commit history and test code co
...
+- A test or suite that covers more than 16,000 files is never skipped.
+- A very small fraction of tests (well under one percent) may run even when they could have been skipped.
```
Comment (OliviaShoup) [inline content/en/tests/test_impact_analysis/troubleshooting/_index.md]: Percentages take numerals and the `%` symbol rather than spelled-out words.

```suggestion
- A very small fraction of tests (well under 1%) may run even when they could have been skipped.
```

File: content/en/tests/test_impact_analysis/_index.md [hunk 12]
```diff
@@ -49,6 +49,26 @@ There are several configuration mechanisms that you can use in these scenarios t
 - If GitHub is your source code management provider, use the `ITR:NoSkip` label (case insensitive) to prevent Test Impact Analysis from skipping tests in pull requests. To use this feature, configure the GitHub App using the [GitHub integration tile][9] with the `Software Delivery: Collect Pull Request Information` feature enabled. This mechanism does not work with tests executed on GitHub actions triggered by `pull_request` events.
 - You can add a list of [excluded branches](#excluded-branches), which disables Test Impact Analysis in those branches.
 
+## Limitations and expected behavior
+
+Test Impact Analysis decides which tests to skip by matching the files each test covers against the files a commit changes. The behaviors below are the known cases where that matching is intentionally conservative, so you know what to expect.
```
Comment (OliviaShoup) [inline content/en/tests/test_impact_analysis/_index.md]: "below" is directional language — the editorial reference calls for "the following" instead, and directional references break when content is re-chunked for AI retrieval.

```suggestion
Test Impact Analysis decides which tests to skip by matching the files each test covers against the files a commit changes. The following behaviors are the known cases where that matching is intentionally conservative, so you know what to expect.
```

File: content/en/tests/test_impact_analysis/_index.md [hunk 13]
```diff
@@ -49,6 +49,26 @@ There are several configuration mechanisms that you can use in these scenarios t
...
+
+To keep analysis fast and predictable, Test Impact Analysis applies a few size limits. Each one is safe: it can only result in more tests running.
+
+- Datadog analyzes up to the 100 most recent commits for a change. On very large changesets, some older commits may not be analyzed, so slightly fewer tests are skipped.
+- If a single commit changes more than 5,000 files, that commit is not analyzed for impact, and its tests are run rather than skipped.
+- If a single test or suite covers more than 16,000 files, it is not skipped.
+
+### Accuracy
+
+To check whether a test covers a changed file efficiently, Datadog uses a space-efficient probabilistic data structure (a Bloom filter). These structures have a small, inherent false-positive rate (approximately 0.04%), so a very small fraction of tests that could have been skipped may run anyway. This never causes a test to be skipped when it should have run.
```
Comment (OliviaShoup) [inline content/en/tests/test_impact_analysis/_index.md]: Misplaced modifier: as written, "efficiently" attaches to "a changed file" rather than to "check." "Efficiently" and "space-efficient" also collide in the same sentence. And the comma in "small, inherent" is wrong — those aren't coordinate adjectives (you wouldn't say "small and inherent").

```suggestion
To check whether a test covers a changed file, Datadog uses a space-efficient probabilistic data structure (a Bloom filter). These structures have a small inherent false-positive rate (approximately 0.04%), so a small fraction of tests that could have been skipped may run anyway.
```

This also clears Vale's `very` warning and shortens a 30-word sentence. I dropped the trailing "This never causes a test to be skipped when it should have run" since the alert above already states that guarantee — restore it if you'd rather keep the reassurance here.

File: content/en/tests/test_impact_analysis/_index.md [hunk 14]
```diff
@@ -49,6 +49,26 @@ There are several configuration mechanisms that you can use in these scenarios t
...
 
+## Limitations and expected behavior
+
+Test Impact Analysis decides which tests to skip by matching the files each test covers against the files a commit changes. The behaviors below are the known cases where that matching is intentionally conservative, so you know what to expect.
+
+<div class="alert alert-info">All of the behaviors in this section are safe by design: they can only cause <strong>more</strong> tests to run, never cause a test to be skipped when it should have run. For code changes that Datadog cannot detect automatically, see <a href="#out-of-the-box-configuration-limitations">Out-of-the-box configuration limitations</a>.</div>
+
+### Scale limits
+
+To keep analysis fast and predictable, Test Impact Analysis applies a few size limits. Each one is safe: it can only result in more tests running.
```
Comment (OliviaShoup) [inline content/en/tests/test_impact_analysis/_index.md]: The second sentence repeats the guarantee already made in the alert directly above. "To keep analysis fast and predictable" is also filler, and "a few size limits" is vaguer than the three specific limits that follow.

```suggestion
Test Impact Analysis applies the following size limits:
```

File: content/en/tests/test_impact_analysis/_index.md [hunk 15]
```diff
@@ -49,6 +49,26 @@ There are several configuration mechanisms that you can use in these scenarios t
...
+
+- Datadog analyzes up to the 100 most recent commits for a change. On very large changesets, some older commits may not be analyzed, so slightly fewer tests are skipped.
```
Comment (OliviaShoup) [inline content/en/tests/test_impact_analysis/_index.md]: Vale flags `very` (`Datadog.words`). "On large changesets" carries the same meaning.

File: content/en/tests/test_impact_analysis/_index.md [hunk 16]
```diff
@@ -49,6 +49,26 @@ There are several configuration mechanisms that you can use in these scenarios t
...
+- If a single commit changes more than 5,000 files, that commit is not analyzed for impact, and its tests are run rather than skipped.
+- If a single test or suite covers more than 16,000 files, it is not skipped.
+
+### Accuracy
```
Comment (OliviaShoup) [inline content/en/tests/test_impact_analysis/_index.md]: `### Accuracy` is a generic heading — a weak retrieval anchor for AI, and it tells a skimmer nothing. Something like `### False positives in coverage matching` names the actual content.

File: content/en/tests/test_impact_analysis/_index.md [hunk 17]
```diff
@@ -111,6 +131,8 @@ Code coverage backfilling adjusts the total reported coverage to include tests o
 
 Code coverage backfilling is supported for Java, .NET, Go, and JavaScript. It is not supported for Ruby, Python, or Swift.
 
+Backfilling is best-effort. For very large test suites, the coverage of some tests skipped by Test Impact Analysis may not be backfilled, so the reported total coverage can be slightly lower than the true coverage. This does not affect which tests are skipped.
```
Comment (OliviaShoup) [inline content/en/tests/test_impact_analysis/_index.md]: 30-word middle sentence, plus a `very` warning from Vale. Splitting it also lets "This does not affect..." get a real subject instead of a bare "This." Note `best-effort` -> `best effort`: compound modifiers generally lose the hyphen in predicate position.

```suggestion
Backfilling is best effort. For large test suites, the coverage of some tests skipped by Test Impact Analysis may not be backfilled. As a result, the reported total coverage can be slightly lower than the true coverage. Backfilling does not affect which tests are skipped.
```

File: content/en/tests/test_impact_analysis/_index.md [hunk 18]
```diff
@@ -49,6 +49,26 @@ There are several configuration mechanisms that you can use in these scenarios t
...
+
+- Datadog analyzes up to the 100 most recent commits for a change. On very large changesets, some older commits may not be analyzed, so slightly fewer tests are skipped.
+- If a single commit changes more than 5,000 files, that commit is not analyzed for impact, and its tests are run rather than skipped.
+- If a single test or suite covers more than 16,000 files, it is not skipped.
+
+### Accuracy
+
+To check whether a test covers a changed file efficiently, Datadog uses a space-efficient probabilistic data structure (a Bloom filter). These structures have a small, inherent false-positive rate (approximately 0.04%), so a very small fraction of tests that could have been skipped may run anyway. This never causes a test to be skipped when it should have run.
+
+If Test Impact Analysis behaves unexpectedly, see the [Troubleshooting][12] page for symptom-based guidance.
```
Comment (OliviaShoup) [inline content/en/tests/test_impact_analysis/_index.md]: "symptom-based guidance" is mildly jargony. "See the [Troubleshooting][12] page" alone does the job, or "for guidance organized by symptom."

**Review by OliviaShoup — CHANGES_REQUESTED** [review]
Thanks for documenting this! The PR looks great. Most of my inline comments are small fixes. Three larger things first.

**1. This removes the only actionable fix from the troubleshooting page.** The rewrite of "Fewer tests skipped than expected" drops both the 50,000-test threshold and the Ruby suite-level skipping workaround (`DD_TESTOPTIMIZATION_TIA_TEST_SKIPPING_MODE=suite`), and neither reappears anywhere. I grepped the English content tree: that env var now survives only in `content/en/tests/test_parallelization/configuration.md` as a bare entry in a `ddtest` config reference table, on a different product's page (not framed as a remedy, and not findable by someone hitting this symptom). The section becomes purely descriptive: it explains why fewer tests were skipped but gives readers nothing to do about it.

Could you confirm: is the 50,000-test limitation obsolete under the new implementation? And does the Ruby workaround still apply? If it does, please restore it. If the new implementation makes it unnecessary, a sentence saying so would help readers who found it before and come looking.

**2. Are these limits gated to the new implementation?** The PR description frames this as a new implementation with new limitations, but the docs present 100 commits / 5,000 files / 16,000 files / 0.04% as flat universal facts with no version or rollout qualifier. If any customers are still on the old implementation, they'll see behavior that doesn't match. Any such constraint should sit immediately under the heading.

**3. Two "limitations" sections now pull in opposite directions.** The page already has "Out-of-the-box configuration limitations," which covers tests being *wrongly skipped* (the unsafe direction). The new section covers *extra tests running* (the safe direction). The info alert disambiguates them nicely, but the heading is what lands in the TOC and anchors AI retrieval, and "Limitations and expected behavior" doesn't signal which kind. Consider "Scale and accuracy limits" or "When more tests run than expected."

Two smaller structural notes: the safety guarantee is stated three times in nine lines (the alert, line 60, and line 68). The alert says it best, so the others can go. And the two pages state the same limits in different words ("the first 100 recent commits" vs. "the 100 most recent commits"; "well under one percent" vs. "approximately 0.04%"), so a reader comparing them can't tell they're the same limit.

Two pre-existing nits outside the diff, FYI only: the alt text at `_index.md:128` reads "ITest Impact Analysis," and the `### Java, .NET, and Go` heading at line 136 trips Vale's sentence-case rule. Fix opportunistically or ignore :)

**Comment by TovRudyy** [comment]
👋 Thanks for the review @OliviaShoup . I reviewed your suggestions, and am answering some unclear topics below. 


> Could you confirm: is the 50,000-test limitation obsolete under the new implementation? And does the Ruby workaround still apply? If it does, please restore it. If the new implementation makes it unnecessary, a sentence saying so would help readers who found it before and come looking.

Yes it is obsolete. Updated.

> **2. Are these limits gated to the new implementation?** The PR description frames this as a new implementation with new limitations, but the docs present 100 commits / 5,000 files / 16,000 files / 0.04% as flat universal facts with no version or rollout qualifier. If any customers are still on the old implementation, they'll see behavior that doesn't match. Any such constraint should sit immediately under the heading.

The limits are applied universally no matter the customer setup. There is no version rollout.

**Review by OliviaShoup — APPROVED** [review]
thanks for addressing all the feedback! looks great 🚀
